freedomofpress/securedrop
# Manage rq services with systemd instead of supervisor

## Status

Ready for review

## Description of Changes

This change makes securedrop-app-code Conflict/Replace supervisor, to ensure that supervisor is purged from the system. The rq worker and requeuing processes are now managed as systemd services, with configuration files in /lib/systemd/system. We're only using rq for submission hashing now. Deletion is handled by a new systemd service, discussed in detail below.

All the services now log to the systemd journal, which in the current SecureDrop server configuration means that their log messages also wind up in /var/log/syslog.

Submission hashing works as before. Requeuing a job interrupted by a reboot is less likely to work, as the worker seems to be shut down such that the running jobs are terminated and moved by rq to the failed queue, but hashing will be done automatically when submissions are downloaded.

Instead of submitting jobs to rq to delete submissions or replies, we'll move files to be deleted to a designated directory in the store, and use a systemd service, securedrop_shredder, to monitor that directory and clear its contents securely.

Moving files is atomic and quick, so should never be interrupted. If the secure deletion is interrupted, the securedrop_shredder service will just find the file on its next pass.

We had discussed adding columns to the database tables involved, and having the service use those to discover items to delete, but: 
- It would have been a bigger, more complicated change.
- It would have introduced the possibility of source information being present in the database for quite a while after the admin "deleted" it, which could be problematic in the event of raid/seizure. This way only the encrypted files might linger; the database records will still be deleted immediately.
- It would have required that we always remember to filter using those columns to avoid working with deleted content. (SQLAlchemy currently has [only partial support for setting up this filtering globally](https://github.com/sqlalchemy/sqlalchemy/wiki/FilteredQuery).

Fixes #4783.

## Testing

Use the staging environment to confirm the asynchronous services still work:

- `make build-debs`
- `make staging`
- Visit your staging source interface and submit a message.
- Confirm that it was hashed by checking the submissions table in the database.
- Verify the hashing operation appears in the `securedrop_rqworker` service log by running `sudo journalctl -u securedrop_rqworker --no-pager` -- the output should look like:
```
-- Logs begin at Fri 2019-09-20 15:20:20 UTC, end at Fri 2019-09-20 18:38:13 UTC. --
Sep 20 18:25:35 app-staging rqworker[1513]: 18:25:35 default: store.queued_add_checksum_for_file(<class 'models.Submission'>, 1, '/var/lib/securedrop/store/LB4IY3SOHQLOZ6FNU4UPALZXRLUOEJKNNLZCNAJXQW2JQZTEQV2TVQLUMBND3Y4VX3QC4K6KDWFCNH2R2YSRZSP7XRPCYO2MSEWOUBY=/1-culinary_fracture-msg.gpg', 'sqlite:////var/lib/securedrop/db.sqlite') (c221907f-5a8b-4676-a835-888a9d321c0b)
Sep 20 18:25:35 app-staging rqworker[1513]: 18:25:35 default: Job OK (c221907f-5a8b-4676-a835-888a9d321c0b)
Sep 20 18:25:35 app-staging rqworker[1513]: 18:25:35 Result is kept for 500 seconds
Sep 20 18:25:35 app-staging rqworker[1513]: 18:25:35 Cleaning registries for queue: default
```
- Add an admin user.
- Visit the journalist interface as that user and:
  - add some replies to the source.
  - delete one. 
  - delete several.
  - delete an entire source collection.

You can watch the securedrop_shredder service with `journalctl -f -u securedrop_shredder --no-pager`. I find it helpful to `sudo apt install tree` on the app server and use that to watch `/var/lib/securedrop`. If you don't want to wait 60 seconds between shredder runs, you can edit `/etc/systemd/system/multi-user.target.wants/securedrop_shredder.service` to change the interval, or just stop the service with `systemctl stop securedrop_shredder`, and then invoke `/var/www/securedrop/scripts/shredder` manually as you delete submissions and replies in the UI. (And yes, you can test the basic functionality in the dev container that way too, if you want to check it out before going to the trouble of building the staging environment.)

## Deployment

With this change, SecureDrop's asynchronous worker processes (the rq worker for hashing submissions, the monitor that requeues those jobs if they're interrupted, and the new service that securely deletes submissions and replies) will be managed by systemd instead of supervisor, and the supervisor package will be purged.

## Checklist

### If you made changes to the server application code:


### If you made changes to the system configuration:


### If you made non-trivial code changes:
